definitions/npm/react-router-dom_v5.x.x/flow_v0.63.x-/react-router-dom_v5.x.x.js

```diff
@@ -0,0 +1,171 @@
+    state?: any
+  };
+
+  declare export type HistoryAction = "PUSH" | "REPLACE" | "POP";
+
+  declare export type RouterHistory = {
+    length: number,
+    location: Location,
+    action: HistoryAction,
+    listen(
+      callback: (location: Location, action: HistoryAction) => void
+    ): () => void,
+    push(path: string | LocationShape, state?: any): void,
+    replace(path: string | LocationShape, state?: any): void,
+    go(n: number): void,
+    goBack(): void,
+    goForward(): void,
+    canGo?: (n: number) => boolean,
+    block(
+      callback: (location: Location, action: HistoryAction) => boolean
+    ): void,
+    // createMemoryHistory
+    index?: number,
+    entries?: Array<Location>
+  };
+
+  declare export type Match = {
+    params: { [key: string]: ?string },
+    isExact: boolean,
+    path: string,
+    url: string
+  };
+
+  declare export type ContextRouter = {|
+    history: RouterHistory,
+    location: Location,
+    match: Match,
+    staticContext?: StaticRouterContext
+  |};
+
+  declare type ContextRouterVoid = {
+    history: RouterHistory | void,
+    location: Location | void,
+    match: Match | void,
+    staticContext?: StaticRouterContext | void
+  };
+
+  declare export type GetUserConfirmation = (
+    message: string,
+    callback: (confirmed: boolean) => void
+  ) => void;
+
+  declare export type StaticRouterContext = {
+    url?: string
+  };
+
+  declare export var StaticRouter: React$ComponentType<{|
+    basename?: string,
+    location?: string | Location,
+    context: StaticRouterContext,
+    children?: React$Node
+  |}>
+
+  declare export var MemoryRouter: React$ComponentType<{|
+    initialEntries?: Array<LocationShape | string>,
+    initialIndex?: number,
+    getUserConfirmation?: GetUserConfirmation,
+    keyLength?: number,
+    children?: React$Node
+  |}>
+
+  declare export var Router: React$ComponentType<{|
+    history: RouterHistory,
+    children?: React$Node
+  |}>
+
+  declare export var Prompt: React$ComponentType<{|
+    message: string | ((location: Location) => string | boolean),
+    when?: boolean
+  |}>
+
+  declare export var Redirect: React$ComponentType<{|
+    to: string | LocationShape,
+    push?: boolean,
+    from?: string,
+    exact?: boolean,
+    strict?: boolean
+  |}>
+
+  declare export var Route: React$ComponentType<{|
+    component?: React$ComponentType<*>,
+    render?: (router: ContextRouter) => React$Node,
+    children?: React$ComponentType<ContextRouter> | React$Node,
+    path?: string | Array<string>,
+    exact?: boolean,
+    strict?: boolean,
+    location?: LocationShape,
+    sensitive?: boolean
+  |}>
+
+  declare export var Switch: React$ComponentType<{|
+    children?: React$Node,
+    location?: Location
+  |}>
+
+  declare export function withRouter<Props : {}, Component: React$ComponentType<Props>>(WrappedComponent: Component) : React$ComponentType<$Diff<React$ElementConfig<$Supertype<Component>>, ContextRouterVoid>>;
```

Comment:
Would be cool to update the HOC typing with [latest recommendations](https://flow.org/en/docs/react/hoc/#toc-hocs-as-of-0-89-0), especially since [`$Supertype` is deprecated](https://github.com/flow-typed/flow-typed/issues/2991).
Maybe in a future PR.

Reply:
As per [[1]](https://flow.org/en/docs/react/hoc/#toc-injecting-props) and [[2]](https://flow.org/en/docs/react/types/#toc-react-abstractcomponent,) I updated the v4 typings to look like so:
```
 declare export function withRouter<Config: {}, Instance>(
    WrappedComponent: React$AbstractComponent<Config, Instance>
  ): React$AbstractComponent<$Diff<Config, ContextRouterVoid>, Instance>;
```
They seem to work fine locally.  Want to drop them into this PR and double-check they pass tests?  I can also do it, I just have never contributed to this repo before :smile:

Reply:
@bezi Would be rad, but maybe wait for this PR to get merged before?

Reply:
https://github.com/flow-typed/flow-typed/pull/3224#discussion_r268489221 leads to test errors for me. when replacing the block with your new definition, i get this error:
```
    Error --------------------------------------------------------------------------------- react-router-dom_v5.x.x.js:157:6
    
    Cannot use type without exactly 3 type arguments.
    
       157|   ): React$AbstractComponent<$Diff<Config, ContextRouterVoid>, Instance>;
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Reply:
Well yeah, let's see this in another MR.